hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/move/MoveManagerImpl.java

```diff
@@ -0,0 +1,750 @@
+      // we just complete the future without sending a delete command.
+      LOG.info("can not remove source replica after successfully " +
+          "replicated to target datanode");
+      completeMove(cid, MoveResult.DELETE_FAIL_POLICY);
+    }
+  }
+
+  /**
+   * whether the given two ContainerPlacementStatus are actually equal.
+   *
+   * @param cps1 ContainerPlacementStatus
+   * @param cps2 ContainerPlacementStatus
+   */
+  private boolean isPlacementStatusActuallyEqual(
+      ContainerPlacementStatus cps1,
+      ContainerPlacementStatus cps2) {
+    return (!cps1.isPolicySatisfied() &&
+        cps1.actualPlacementCount() == cps2.actualPlacementCount()) ||
+        cps1.isPolicySatisfied() && cps2.isPolicySatisfied();
+  }
+
+  /**
+   * Returns whether {Existing replicas + Target_Dn - Source_Dn}
+   * satisfies current placement policy.
+   * @param srcDn DatanodeDetails of source data node
+   * @param targetDn DatanodeDetails of target data node
+   * @param cInfo container replicas
+   * @return whether the placement policy is satisfied after move
+   */
+  private boolean isPolicySatisfiedAfterMove(
+      ContainerInfo cInfo, DatanodeDetails srcDn,
+      DatanodeDetails targetDn) throws ContainerNotFoundException {
+    ContainerID cid = cInfo.containerID();
+    final Set<ContainerReplica> replicas =
+        containerManager.getContainerReplicas(cid);
+    Set<ContainerReplica> movedReplicas =
+        replicas.stream().collect(Collectors.toSet());
+    movedReplicas.removeIf(r -> r.getDatanodeDetails().equals(srcDn));
+    movedReplicas.add(ContainerReplica.newBuilder()
+        .setDatanodeDetails(targetDn)
+        .setContainerID(cid)
+        .setContainerState(StorageContainerDatanodeProtocolProtos
+            .ContainerReplicaProto.State.CLOSED).build());
+    ContainerPlacementStatus placementStatus = getPlacementStatus(
+        movedReplicas, cInfo);
+    return placementStatus.isPolicySatisfied();
+  }
+
+  /**
+   * Given a set of ContainerReplica, transform it to a list of DatanodeDetails
+   * and then check if the list meets the container placement policy.
+   * @param replicas List of containerReplica
+   * @param cInfo container info
+   * @return ContainerPlacementStatus indicating if the policy is met or not
+   */
+  private ContainerPlacementStatus getPlacementStatus(
+      Set<ContainerReplica> replicas, ContainerInfo cInfo) {
+    List<DatanodeDetails> replicaDns = replicas.stream()
+        .map(ContainerReplica::getDatanodeDetails)
+        .collect(Collectors.toList());
+    return placementPolicyValidateProxy.validateContainerPlacement(
+        replicaDns, cInfo);
+  }
+
+  /**
+   * Sends replicate container command for the given container to the given
+   * datanode.
+   *
+   * @param cid Container id to be replicated
+   * @param tgt The destination datanode to replicate
+   * @param src List of source nodes from where we can replicate
+   */
+  private void sendReplicateCommand(
```

Comment:
We should probably have a common method on ReplicationManager to send the command. At the moment there are a few places in RM that create and send commands, so it would be good to bring them together into a single place, as there are a few things that need to be taken care of together, which are hard to check when the logic is split into various places.

1. Updating pending ops.
2. Adding the SCM term.
3. Ensuring we add the correct replica index.

We should refactor the current RM calls into a single place in RM in another Jira and then we can use the same interface here.

Reply:
make sense , we can create another jira to handle this

Reply:
https://github.com/apache/ozone/pull/3990 creates a common point to send commands to the DNs via the RM. Once it is committed we can use it here.

Reply:
since #3990 is merged,   let us use it